File: app/src/main/java/com/arturogutierrez/openticator/domain/account/add/AddAccountFormPresenter.java
package com.arturogutierrez.openticator.domain.account.add;

import com.arturogutierrez.openticator.domain.account.interactor.AddAccountInteractor;
import com.arturogutierrez.openticator.domain.account.model.Account;
import com.arturogutierrez.openticator.interactor.DefaultSubscriber;
import com.arturogutierrez.openticator.view.presenter.Presenter;
import javax.inject.Inject;

public class AddAccountFormPresenter extends DefaultSubscriber<Account> implements Presenter {

  private final AddAccountInteractor addAccountInteractorInteractor;
  private AddAccountView view;

  @Inject
  public AddAccountFormPresenter(AddAccountInteractor addAccountInteractorInteractor) {
    this.addAccountInteractorInteractor = addAccountInteractorInteractor;
  }

  public void setView(AddAccountView view) {
    this.view = view;
  }

  @Override
  public void resume() {

  }

  @Override
  public void pause() {

  }

  @Override
  public void destroy() {
    addAccountInteractorInteractor.unsubscribe();
  }

  public void createTimeBasedAccount(String accountName, String accountSecret) {
    if (accountName == null || accountName.trim().length() == 0) {
      view.showFieldError(AddAccountView.FieldError.NAME);
    } else if (accountSecret == null || accountSecret.trim().length() == 0) {
      // TODO: Add validation to support only base 32 chars
      view.showFieldError(AddAccountView.FieldError.SECRET);
    } else {
      addAccountInteractorInteractor.configure(accountName.trim(), accountSecret.trim());
      addAccountInteractorInteractor.execute(this);
    }
  }

  @Override
  public void onNext(Account account) {
    view.dismissForm();
  }

  @Override
  public void onError(Throwable e) {
    view.unableToAddAccount();
  }
}

File: app/src/main/kotlin/com/arturogutierrez/openticator/domain/account/add/AddAccountFormPresenter.kt
package com.arturogutierrez.openticator.domain.account.add

import com.arturogutierrez.openticator.domain.account.AccountFactory
import com.arturogutierrez.openticator.domain.account.add.AddAccountView.InputError.EMPTY_ACCOUNT_NAME
import com.arturogutierrez.openticator.domain.account.add.AddAccountView.InputError.EMPTY_SECRET
import com.arturogutierrez.openticator.domain.account.add.AddAccountView.InputError.INVALID_SECRET
import com.arturogutierrez.openticator.domain.account.interactor.AddAccountInteractor
import com.arturogutierrez.openticator.domain.account.interactor.AddAccountInteractor.Params
import com.arturogutierrez.openticator.domain.account.model.Account
import com.arturogutierrez.openticator.helpers.isBase32
import com.arturogutierrez.openticator.interactor.DefaultSingleObserver
import com.arturogutierrez.openticator.view.presenter.Presenter
import javax.inject.Inject

class AddAccountFormPresenter @Inject constructor(private val accountFactory: AccountFactory,
                                                  private val addAccountInteractorInteractor: AddAccountInteractor) : Presenter<AddAccountView>() {

  override fun destroy() {
    addAccountInteractorInteractor.dispose()
  }

  fun createTimeBasedAccount(accountName: String?, accountSecret: String?) {
    if (accountName == null || accountName.trim().isEmpty()) {
      view?.showFieldError(EMPTY_ACCOUNT_NAME)
    } else if (accountSecret == null || accountSecret.trim().isEmpty()) {
      view?.showFieldError(EMPTY_SECRET)
    } else if (!accountSecret.isBase32()) {
      view?.showFieldError(INVALID_SECRET)
    } else {
      val newAccount = accountFactory.createAccount(accountName.trim(), accountSecret.trim())
      val params = Params(newAccount)
      addAccountInteractorInteractor.execute(params, object : DefaultSingleObserver<Account>() {
        override fun onSuccess(t: Account) {
          onAccountAdded()
        }

        override fun onError(e: Throwable) {
          onErrorAddingAccount()
        }
      })
    }
  }

  private fun onAccountAdded() {
    view?.dismissForm()
  }

  private fun onErrorAddingAccount() {
    view?.unableToAddAccount()
  }
}
